### ISSUE TITLE
Grids don't render correctly

### ISSUE THREAD

**Issue by JonathanMandel** [jwstegemann/fritz2#831]
**Describe the bug**
I've written a web app that uses Tailwind CSS and Fritz2 1.0 RC11. It uses Tailwind grid commands to generate a spreadsheet-style grid. However, I've found that properly coded grids do not render correctly sometimes. 

Here is the code that does part of the work. It renders the column headings as a 1-row grid.

``
fun RenderContext.columnHeadingFramework(
    parms:     Store<ColumnHeadingFramworkParameterPack>
)
{
    // 4-column version
    div ("grid grid-cols-4 w-full gap-2 mt-0") {
        div("col-span-3") {}
        div("col-span-1") {}
    }

    // 6-column version
    div ("grid grid-cols-6 w-full gap-2 mt-0") {
        div("col-span-3") {}
        div("col-span-2") {}
        div("col-span-1") {}
    }

    // 10-column version
    div ("grid grid-cols-10 w-full gap-2 mt-0") {
        div("col-span-4") {}
        div("col-span-2") {}
        div("col-span-2") {}
        div("col-span-1") {}
        div("col-span-1") {}
    }

    parms.data.render { p ->
        val gridCols = p.gridSpans.sum()
        div("w-full isolate grid grid-cols-$gridCols w-full gap-2 mt-6") {
            //div("") {
                for (i in p.labels.indices) {
                    val label: String
                    val labelFormat: String
                    if (p.labels[i].first() == '*') {
                        labelFormat = "justify-self-center uppercase text-gray-700 text-xs font-bold"
                        label = p.labels[i].drop(1)
                    } else {
                        labelFormat = "px-4 uppercase text-gray-700 text-xs font-bold"
                        label = p.labels[i]
                    }
                    console.log("columnHeadingFramework label: $label, col-span = ${p.gridSpans[i]}")
                    div("col-span-${p.gridSpans[i]}") {
                        val s = p.tips.getOrNull(i)
                        if (s != null) {
                            label(labelFormat) {
                                `for`("grid-col-$i")
                                +label
                            }.tooltip(UI.tooltipStyle){
                                +s
                                placement = PlacementValues.top
                                arrow()
                            }
                        } else {
                            label(labelFormat) {
                                `for`("grid-col-$i")
                                +label
                            }
                        }
                    }
                }
            //}
        }
    }
}
``

![Broken final table](https://github.com/jwstegemann/fritz2/assets/3219814/1e27ecf3-925b-4804-bd6a-b9fd40b18b3f)
![Not broken final table](https://github.com/jwstegemann/fritz2/assets/3219814/145a7655-416a-4d04-a5a2-b70b36090e51)

These two images show what it's supposed to look like when it breaks and how it looks when it works.

To get the code to fail, comment out the empty div()s at the top of the function. 

Now here's the odd thing. It doesn't always fail as soon as you build without those empty div()s. It may work for a few builds, then it fails. Putting the empty div()s back in reliably fixes the problem.  The empty div()s match the grid patterns that get passed into this function. 

I have tried every variant of this function I can imagine, including changing the composition of Store objects I pass into the function, and passing plain data (no Store). I've also tried a bare-bones version of the function, which exhibits the same behavior.

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Browser: Chrome Version 120.0.6099.111 (Official Build) (64-bit)
 - Browser: Firefox Version 120.0 (64-bit)
 - Fritz2 vRC11

**Additional context**
You can see a recent working version of this app at www.bakersformulary.com.

**Comment by haukesomm** [comment]
Hi @JonathanMandel, thanks for reaching out with your issue!

In your code you are using String interpolation to dynamically construct Tailwind classes like this: `col-span-${p.gridSpans[i]}`.

Tailwind class names must be present at compile time in order to be recognized and not be optimized out. By using String interpolation the final class name is only present at runtime which may result in the class not being part of your stylesheet depending on your configuration.

Does the problem persist if you replace your dynamically generated class names with static class names (e.g. `grid-cols-4`)? If so, your issue is probably caused by the mechanics described above.

**Comment by JonathanMandel** [comment]
Thank you for replying.

What you say makes sense. And there are two things that I don't understand.

First, in your own example code, you use string composition. Here's an
example from TicTacToe:

div("row") {
    gameStore.field.renderEach { cell ->
           div("col-4 border d-flex justify-content-center
align-items-center") {
                inlineStyle("height: 10rem; background-color: ${if
(cell.isInWinningGroup) "green" else "white"};")

Please explain why this code works if "background-color: green" and
"background-color: white" haven't been hardcoded?

Second, if there is no way to alter Tailwind formatting under program
control, Tailwind is totally useless in production coding environments. Yet
Tailwind is one of the most popular CSS libraries in existence. How are
formats composed programmatically in the real world?

Thanks.

**Comment by Lysander** [comment]
You quoted an example that works with *Inline styles*, so no tailwindcss there, just plain old CSS!

We should rework that imho, as this could be solved in a better way without inline styles and obviously disturbs users.

You simply must use Kotlins control flow mechanisms like `if...else` or `when` to apply the classes that are needed on certain states.

Iirc there are examples in the documentation in the rendering chapter that explicitly deals with reactive styling.

**Comment by JonathanMandel** [comment]
Got it. So, can I use string composition, as long as the classes that might
get composed are hardcoded elsewhere? For example, if I use "colspan-$cv",
and I also (somewhere) have "colspan-3" and whatever other classes I
generate? In other words, if I defeat the optimization by naming the
classes I'm going to generate programmatically?

If I am required to use Kotlin control flow to generate Tailwind variants,
my code is going to balloon considerably. "colspan-$cv" becomes

val fmt = when(cv) {
1 -> "colspan-1"
2 -> "colspan-2"
3 -> "colspan-3"
4 -> "colspan-4"
5 -> "colspan-5"
6 -> "colspan-6"
7 -> "colspan-7"
8 -> "colspan-8"
9 -> "colspan-9"
10 -> "colspan-10"
else -> "colspan-0"
}

or an array-based variant. Yes, I can hide those all in a support file, but
is this actually the only way to do this?

**Comment by haukesomm** [comment]
Exactly, this is the only way I could think of at the moment. I'm not sure if a more dynamic approach is easily doable at the moment.

**Comment by Lysander** [comment]
> For example, if I use "colspan-$cv", and I also (somewhere) have "colspan-3" and whatever other classes I generate? In other words, if I defeat the optimization by naming the classes I'm going to generate programmatically? 

You might use this pattern, but be cautious: The code that contains the tailwind class names must really be "useful" - if not, it will be deleted by the [DCE](https://kotlinlang.org/docs/javascript-dce.html). I would encourage you to *not* use `String`-interpolation to create CSS classes. It resembles premature optimization; read below.

> If I am required to use Kotlin control flow to generate Tailwind variants, my code is going to balloon considerably. "colspan-$cv" becomes val fmt = when(cv) { 1 -> "colspan-1" 2 -> "colspan-2" 3 -> "colspan-3" 4 -> "colspan-4" 5 -> "colspan-5" 6 -> "colspan-6" 7 -> "colspan-7" 8 -> "colspan-8" 9 -> "colspan-9" 10 -> "colspan-10" else -> "colspan-0" } or an array-based variant. Yes, I can hide those all in a support file, but is this actually the only way to do this?

Do not have fear to integrate such class assigning code! We have done this quite often yet within some internal UI-component library. We believe this is no harm at all, but helps to keeps the UI code rather clean of `String`-"magic". Just put such code into a function; we have used this pattern as extension function of some `enum class` types quite often to assign each value its own color-class.

To recap this issue: It is not an issue with fritz2 imho, so you might close it, if there is nothing we could do.

**Comment by JonathanMandel** [comment]
You are right, this is not a problem with Fritz2. Thank you for taking the
time to explain this. I will close the issue.
